### ISSUE TITLE
Fix all IE8 global recursion bugs (e.g. mergePackageBlock() and requireAll())

### ISSUE THREAD

**Issue by tobycatlin** [BladeRunnerJS/brjs#1190]
A bug in IE8 means that IE won't support any execution stack longer than 12 frames.
All is explained in this blog post: http://www.cappuccino-project.org/blog/2010/03/internet-explorer-global-variables-and-stack-overflows.html

Possibly related to this call: https://github.com/BladeRunnerJS/topiarist/issues/11

The following will trigger the error:
```javascript
var recurse = function(times)
{
    if (times !== 0)
        recurse(times - 1);
}

recurse(13);
```

In mergePackageBlock there is a similar recursive call:

```javascript
window.mergePackageBlock = function(context, packageBlock) {
		for (var packageName in packageBlock) {
			if(!context[packageName]) {
				context[packageName] = packageBlock[packageName];
			}
			else {
				mergePackageBlock(context[packageName], packageBlock[packageName]);
			}
		}
	}
```
http://yourls.caplin.com/c

**Comment by tobycatlin** [comment]
Customer also mentions requireAll may trigger the error but only because it is on window. Not sure if it is also vunerable.

**Comment by dchambers** [comment]
> A bug in IE8 means that IE won't support any execution stack longer than 12 frames.

Not quite @tobycatlin, if that were true almost no code would run at all!

Instead, from reading the resource you link, I can see that it only happens for functions that are assigned to the global object (i.e. `window`) from the global scope, so that converting this:

```js
window.recurse = function(times)
{
    if (times !== 0)
        recurse(times - 1);
}

recurse(13);
```

to this:

```js
var recurse = function(times)
{
    if (times !== 0)
        recurse(times - 1);
}

recurse(13);
```

or this

```js
(function(){
    var recurse = function(times)
    {
        if (times !== 0)
            recurse(times - 1);
    }

    //we won't have access outside to recurse, so add a global ref
    window.recurse = recurse;
})();

recurse(13);
```

fixes the problem. So it's actually a really obscure edge-case, that's also really easy to fix once you've diagnosed it.

**Comment by dchambers** [comment]
> Customer also mentions requireAll may trigger the error but only because it is on window. Not sure if it is also vunerable.

Yes, we should do a global audit of all of our code and fix any other _latent_ bugs caused by the same underlying issue. I'm therefore going to rename this issue.

**Comment by ioanalianabalas** [comment]
Fixed by #1194.
